core/src/main/java/org/elasticsearch/index/search/NestedHelper.java

```diff
@@ -0,0 +1,161 @@
+            return false;
+        }
+        for (String parent = parentObject(field); parent != null; parent = parentObject(parent)) {
+            ObjectMapper mapper = mapperService.getObjectMapper(parent);
+            if (mapper != null && mapper.nested().isNested()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /** Returns true if the given query might match parent documents. */
+    public boolean mightMatchNonNestedDocs(Query query, String nestedPath) {
+        if (query instanceof ConstantScoreQuery) {
+            return mightMatchNonNestedDocs(((ConstantScoreQuery) query).getQuery(), nestedPath);
+        } else if (query instanceof BoostQuery) {
+            return mightMatchNonNestedDocs(((BoostQuery) query).getQuery(), nestedPath);
+        } else if (query instanceof MatchAllDocsQuery) {
+            return true;
+        } else if (query instanceof MatchNoDocsQuery) {
+            return false;
+        } else if (query instanceof TermQuery) {
+            return mightMatchNonNestedDocs(((TermQuery) query).getTerm().field(), nestedPath);
+        } else if (query instanceof PointRangeQuery) {
+            return mightMatchNonNestedDocs(((PointRangeQuery) query).getField(), nestedPath);
+        } else if (query instanceof BooleanQuery) {
+            BooleanQuery bq = (BooleanQuery) query;            
+            boolean hasRequiredClauses = bq.clauses().stream().anyMatch(BooleanClause::isRequired);
+            if (hasRequiredClauses) {
+                return bq.clauses().stream()
+                        .filter(BooleanClause::isRequired)
+                        .map(BooleanClause::getQuery)
+                        .allMatch(q -> mightMatchNonNestedDocs(q, nestedPath));
+            } else {
+                return bq.clauses().stream()
+                        .filter(c -> c.getOccur() == Occur.SHOULD)
+                        .map(BooleanClause::getQuery)
+                        .anyMatch(q -> mightMatchNonNestedDocs(q, nestedPath));
+            }
+        } else {
+            return true;
+        }
+    }
+
+    /** Returns true if a query on the given field might match nested documents. */
+    boolean mightMatchNonNestedDocs(String field, String nestedPath) {
+        if (mapperService.fullName(field) == null) {
+            return false;
+        }
+        for (String parent = parentObject(field); parent != null; parent = parentObject(parent)) {
+            ObjectMapper mapper = mapperService.getObjectMapper(parent);
+            if (mapper!= null && mapper.nested().isNested()) {
+                if (mapper.fullPath().equals(nestedPath)) {
+                    // If the mapper does not include in its parent or in the root object then
+                    // the query might only match nested documents with the given path
+                    return mapper.nested().isIncludeInParent() || mapper.nested().isIncludeInRoot();
```

Comment:
I think copy_to should be fine since it is a different field name. `include_in_parent` and `include_in_root` make things more complicated because the same field name might have documents that belong to multiple spaces.